adambushman/public-library-web-app
# Adam/authentication authorization

## Description

I implemented full auth for the application. It's comprised of two parts:

#### Authentication

This is implemented in all sensitive pages with a check for existence of `isset($_SESSION['accountId'])`. If that doesn't exist, the user is redirected to `login.php`. 

The login page accepts username and password and compares each to the database. Given matching records, a session is started with the value of `accountId`. User is redirected to `home.php`. If records do not match, return to "login" (should have a message display at some point).

It's also present in the `navbar.php` partial, hiding or showing links depending on log in status.

#### Authorization

This is implemented in most pages, even those with links to sensitive pages. It will query the database using the `accountId` saved in the session and pull role or account type. That is then checked against "Admin"/"Staff" roles. Where exists a match, a user may view a sensitive page or interact with links to sensitive pages/forms. 

Where there exists no match (an account with just a "Member" role), they will be redirected to `401.php`, explaining their account lacks access and suggesting other places in the app to which they should navigate.

#### Files

The big pieces to this workflow are in `views/pages/login.php` (the form and inputs), `controllers/authentication/login-controller.php` (the code to confirm credentials and start session), and `controllers/authentication/logout-controller.php` (code to remove/reset session contents or "logging out the user").

### How to test

Launch the app with WAMP. Explore these scenarios:

1. Without logging in, jump around the site via available links. Next, using the URL, try and navigate to "sensitive" pages, like `view-account.php` or `edit-item.php?itemId=1` or `view-employees.php`. You should be redirected to `login.php`. Check to make sure.
2. Login with the "pjones" username and password. Jump around the site via available links. Notably, you should be able to navigate to `view-account.php` via the nav bar and URL. Try to visit staff only pages using the URL like `view-employees.php` or `edit-item.php?itemId=1`. You should get redirected to the `401.php` page.
3. Login with the "bsmith" username and password. Jump around the site via available links. Check they can also access `view-account.php`. Now make sure they can visit and see `view-employees.php` or `edit-item.php?itemId=1`. There should be no redirect.

Let me know if you see something wrong.
